Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,448 @@
+  `4 * (5 + 6)` to the right: `2 * 3 + 4 * 7 + 10 * 10 + 4 * (5 + 6)`.
+
+For the technical description, look at `Mathlib.MoveAdd.weight` and `Mathlib.MoveAdd.reorderUsing`.
+
+`move_add` is the specialization of a more general `move_oper` tactic that takes a binary,
+associative, commutative operation and a list of "operand atoms" and rearranges the operation.
+
+## Extension notes
+
+To work with a general associative, commutative binary operation, `move_oper`
+needs to have inbuilt the lemmas asserting the analogues of
+`add_comm, add_assoc, add_left_comm` for the new operation.
+Currently, `move_oper` supports `HAdd.hAdd`, `HAdd.hAdd`, `and`, `or`, `max`, `min`.
+
+These lemmas should be added to `Mathlib.MoveAdd.move_oper_simpCtx`.
+
+## Implementation notes
+
+The tactic takes the target, replaces the maximal subexpressions whose head symbol is the given
+operation and replaces them by their reordered versions.
+Once that is done, it tries to replace the initial goal with the permuted one by using `simp`.
+
+Currently, no attempt is made at guiding `simp` by doing a `congr`-like destruction of the goal.
+This will be the content of a later PR.
+-/
+
+open Lean Expr
+
+/-- `getExprInputs e` inspects the outermost constructor of `e` and returns the array of all the
+arguments to that constructor that are themselves `Expr`essions. -/
+def Lean.Expr.getExprInputs : Expr → Array Expr
+  | app fn arg        => #[fn, arg]
+  | lam _ bt bb _     => #[bt, bb]
+  | forallE _ bt bb _ => #[bt, bb]
+  | letE _ t v b _    => #[t, v, b]
+  | mdata _ e         => #[e]
+  | proj _ _ e        => #[e]
+  | _ => #[]
+
+/-- `size e` returns the number of subexpressions of `e`. -/
+partial
+def Lean.Expr.size (e : Expr) : ℕ := (e.getExprInputs.map size).foldl (· + ·) 1
+
+namespace Mathlib.MoveAdd
+
+section ExprProcessing
+
+section reorder
+variable {α :Type*} [BEq α]
+
+/-!
+##  Reordering the variables
+
+This section produces the permutations of the variables for `move_add`.
+
+The user controls the final order by passing a list of terms to the tactic.
+Each term can be preceded by `←` or not.
+In the final ordering,
+* terms preceded by `←` appear first,
+* terms not preceded by `←` appear last,
+* all remaining terms remain in their current relative order.
+-/
+
+/-- `uniquify L` takes a list `L : List α` as input and it returns a list `L' : List (α × ℕ)`.
+The two lists `L` and `L'.map Prod.fst` coincide.
+The second component of each entry `(a, n)` in `L'` is the number of times that `a` appears in `L`
+before the current location.
+
+The resulting list of pairs has no duplicates.
+-/
+def uniquify : List α → List (α × ℕ)
+  | []    => []
+  | m::ms =>
+    let lms := uniquify ms
+    (m, 0) :: (lms.map fun (x, n) => if x == m then (x, n + 1) else (x, n))
+
+variable [Inhabited α]
+
+/--  `weight L` takes a list of pairs `(a, h?)` where
+* `a` is a term of some type `α` and
+* `h?` is a `Bool`ean
+
+and returns a function `α → ℤ`.
```

Comment:
The types are already clear so let's try to write a one-line summary:
```suggestion
/-- Return a sorting key so that all `(a, true)`s are in the list's order
and sorted before all `(a, false)`s, which are also in the list's order.
```